Mathlib/Data/Set/Subset.lean

```diff
@@ -0,0 +1,168 @@
+/-
+Copyright (c) 2024 Miguel Marco. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Miguel Marco
+-/
+import Mathlib.Data.Set.Function
+import Mathlib.Data.Set.Functor
+
+/-!
+# Sets in subtypes
+
+This file is about sets in `Set A` when `A` is a set.
+
+It defines notation `↓∩` for sets in a type pulled down to sets in a subtype, as an inverse
+operation to the coercion that lifts sets in a subtype up to sets in the ambient type.
+
+This module also provides lemmas for `↓∩` and this coercion.
+
+## Notation
+
+Let `α` be a `Type`, `A B : Set α` two sets in `α`, and `C : Set A` a set in the subtype `↑A`.
+
+- `A ↓∩ B` denotes `(Subtype.val ⁻¹' B : Set A)` (that is, `{x : ↑A | ↑x ∈ B}`).
+- `↑C` denotes `Subtype.val '' C` (that is, `{x : α | ∃ y ∈ C, ↑y = x}`).
+
+This notation, (together with the `↑` notation for `Set.CoeHead` in `Mathlib.Data.Set.Functor`)
+is scoped to the `Set.Notation` namespace.
+To enable it, use `open Set.Notation`.
+
+
+## Naming conventions
+
+Theorem names refer to `↓∩` as `preimage_val`.
+
+## Tags
+
+subsets
+-/
+
+open Set
+
+variable {ι : Sort*} {α : Type*} {A B C : Set α} {D E : Set A}
+variable {S : Set (Set α)} {T : Set (Set A)} {s : ι → Set α} {t : ι → Set A}
+
+namespace Set.Notation
+
+/--
+Given two sets `A` and `B`, `A ↓∩ B` denotes the intersection of `A` and `B` as a set in `Set A`.
+
+The notation is short for `((↑) ⁻¹' B : Set A)`, while giving hints to the elaborator
+that both `A` and `B` are terms of `Set α` for the same `α`.
+This set is the same as `{x : ↑A | ↑x ∈ B}`.
+-/
+scoped notation3 A:67 " ↓∩ " B:67 => (Subtype.val ⁻¹' (B : type_of% A) : Set (A : Set _))
+
+end Set.Notation
+
+namespace Set
+
+open Notation
+
+lemma preimage_val_eq_univ_of_subset (h : A ⊆ B) : A ↓∩ B = univ := by
+  rw [eq_univ_iff_forall, Subtype.forall]
+  exact h
+
+lemma preimage_val_subset_preimage_val_iff : A ↓∩ B ⊆ A ↓∩ C ↔ A ∩ B ⊆ A ∩ C := by
```

Comment:
I'd argue we should state this as
```suggestion
lemma preimage_val_subset_preimage_val_iff : A ↓∩ B ⊆ A ↓∩ C ↔ B ∩ A ⊆ C ∩ A := by
```
because this matches the direction of `Subtype.image_preimage_coe`.
Alternatively, we could flip `Subtype.image_preimage_coe` and related lemmas to match.

Reply:
In my initial implementation, I actually used the notation `B ↓∩ A` (as in meaning, "`B` is pulled down to `A`), and hence, this kind of results matched other existing results. With that notation, it totally made sense to use the expression you mention.

By @winstonyin suggestion I changed the notation, and then I am not so sure (It is kind of nice to "just change  `B ↓∩ A` to  `B ∩ A` ").

Reply:
I think whatever the notation, [`Subtype.preimage_val_eq_preimage_val_iff`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Set/Image.html#Subtype.preimage_val_eq_preimage_val_iff) should be changed if needed to have the sensible order (I don't think `A ↓∩ B ⊆ A ↓∩ C ↔ B ∩ A ⊆ C ∩ A` is sensible). I would prefer `s ∩ t = s ∩ t` because prefix operators are more common than postfix operators, and I'm seeing `(s ∩ .)` as a prefix operator.

I had suggested the current meaning of `A ↓∩ B`. Here are the options:

* Use `A ↓∩ B` to mean "an intersection down to the subtype for `A`". I suggested this because I am seeing this as a prefix operator `(A ↓∩ .)` where `A` is configuration for the operator (it sets the type of the result too).

* Use `B ↓∩ A` to mean "an intersection with `A` where `B` is pulled down to `A`". If this notation is like a type-ascription-based-coercion, then it has some parallel with `(. : Set A)` for the inverse operation, assuming that would trigger a coercion (though prefix arrow is also the inverse operation). (This is a matter of taste, but I would find the arrow on the other side to be easier to read, as in `B ∩↓ A`, but again, that's just personal taste.)

Reply:
I do agree that the down arrow is better placed in the side of the set that acts as subtype (that is, the place where you are "going down" to.

Your argument about prefearing a prefix operator makes sense.... but then we get that `preimage_val_subset_preimage_val_iff` will either have a weird change of order in the statment, or go in a different direction that `Subtype.image_preimage_coe`.

So, I see the following possible options:

1. Leave it as it is: a prefix operator, with statements that respect the sides of the intersections. The bad part is that it doesn't match `Subtype.image_preimage_coe`.
2. Change the statement so it matches `Subtype.image_preimage_coe`, as @eric-wieser  suggests. The problem with that option is that we get the weird change of order in the operands in the statement.
3. Change the notation to `B ∩↓ A`. That solves both the inconsistencies with  `Subtype.image_preimage_coe` and the order of operands. The problem here would be it is a postfix operator.
4. Change the definition of `Subtype.image_preimage_coe`. That could be quite messy, since many results in mathlib depend on it.

My vote would go to 1 or 3. I personally don't like much 2. and 4. means opening a can of worms that i would rather not deal with.

Reply:
For what it's worth, I don't see any reason `Subtype.image_preimage_coe` needs to be in its current form, and changing it (later) to `((↑) : s → α) '' (((↑) : s → α) ⁻¹' t) = s ∩ t` seems perfectly fine (and maybe better?). I'm through half of mathlib so far testing out the changes.

Reply:
#10433 does this change.

Reply:
So, since some proofs in this PR would be affected by that change, shall we include that PR as a dependency for this one? Or just go on and make the change later?

Reply:
Now that #10433 is merged, this lemma probably belongs next to `Subtype.preimage_val_eq_preimage_val_iff` in the `Subtype` namespace (and in the other file).

Reply:
What is the right way to handle the merging of a branch that this depends on? merging main branch into this one? rebase?

Reply:
Merging the `master` branch should work

Reply:
I moved that lemma to the `Image` file.